### ISSUE TITLE
Typechecker doesn't honor Deref as a trait bound on references

### ISSUE THREAD

**Issue by djrenren** [rust-lang/rust#75208]
I tried this code:

```rust
// Ought to work fine but doesn't
fn borrow_and_deref<T>(p: T) -> i32 where for<'a> &'a T: Deref<Target=i32> {
    *&p
}
```

I expected to see this happen: Successful typechecking

Instead, this happened: 
<img width="723" alt="image" src="https://user-images.githubusercontent.com/634666/89491662-c12c2100-d77d-11ea-990a-81123c62194b.png">

It seems the typechecker can't see that the Deref trait constrains the type of `*&p`. The typechecker thinks that we're trying to return `T` but we know that actually this is returning `i32`. While you can get the same error without the HRTB, it's easier to demonstrate that the trait bound should definitely apply to our borrow of p, since it applies to all of them. 

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9839adeaefe88250d5df73fce68be2ac)

**Comment by djrenren** [comment]
Further reading up on the dereferencing operator makes it seem like this is intended (or at least specified) behavior, because `*x` is only semantically equivalent to `*Deref::deref(&x)` for non-pointer types ([ref](https://doc.rust-lang.org/reference/expressions/operator-expr.html#the-dereference-operator)).  but since there's [an impl for `&'_ [T]`](https://doc.rust-lang.org/std/ops/trait.Deref.html#impl-Deref-5) it seems odd to make this exclusion.

Also, even if there's a specialized implementation of dereferencing, allowing this program to typecheck isn't a soundness issue, since the specialized implementation still adheres to the invariants.

**Comment by djrenren** [comment]
Found the culprit: https://github.com/rust-lang/rust/blob/39d5a61f2e4e237123837f5162cc275c2fd7e625/src/librustc_typeck/check/place_op.rs#L21-L23

It seems we could solve this problem by resolving the impl of Deref first, then emitting that type (with no adjustments to the IR). However, maybe this is just too expensive and not worth it. Thoughts?

**Comment by nbdd0121** [comment]
Sound plausible, but as you said this could potentially be expensive. I am not sure if this would needs a lang-team approval though. cc @nikomatsakis 

@rustbot modify labels: +A-traits +T-lang

**Comment by nikomatsakis** [comment]
Yeah, it sounds plausible indeed, but it may have implications. I wouldn't be opposed to experimenting but I admit I also don't feel very highly motivated, it'd be nice to have a more concrete motivation for this change. Is this just experimentation with strange programs, or is there a use case here?
